javatests/arcs/android/storage/database/DatabaseImplTest.kt

```diff
@@ -2837,6 +2837,525 @@ class DatabaseImplTest {
+                    VersionMap("actor" to 2)
+                )
+            ),
+            schema = POST_UPGRADE_CONTAINER_SCHEMA,
+            databaseVersion = 2,
+            versionMap = VersionMap("actor" to 2)
+        )
+
+        // Note that in this case the returned inline entity doesn't contain a null
+        // value for the new field number; though by the time it's returned it does
+        // report as being of the new schema type.
+        val outputVersionOfEntity1 = DatabaseData.Entity(
+            RawEntity(
+                "entity1",
+                mapOf("inline" to inlineEntity1),
+                emptyMap()
+            ),
+            POST_UPGRADE_CONTAINER_SCHEMA,
+            FIRST_VERSION_NUMBER,
+            VERSION_MAP
+        )
+
+        database.insertOrUpdate(backingKey2, entity2)
+        database.insertOrUpdate(collectionKey, collection2)
+        val outputCollection2 = database.getCollection(collectionKey, POST_UPGRADE_CONTAINER_SCHEMA)
+        assertThat(outputCollection2!!.values.map {
+            database.getEntity(it.reference.storageKey, POST_UPGRADE_CONTAINER_SCHEMA)
+        }).containsExactly(outputVersionOfEntity1, entity2)
+    }
+
+    @Test
+    fun canUpgradeInlineSchema_byAddingNewFields_noWriteCase() = runBlockingTest {
+        val collectionKey = DummyStorageKey("collection")
+        val backingKey1 = DummyStorageKey("backing1")
+
+        val inlineEntity1 = RawEntity(
+            "",
+            mapOf("text" to "forty two".toReferencable()),
+            emptyMap()
+        )
+
+        val entity1 = DatabaseData.Entity(
+            RawEntity(
+                "entity1",
+                mapOf("inline" to inlineEntity1),
+                emptyMap()
+            ),
+            PRE_UPGRADE_CONTAINER_SCHEMA,
+            FIRST_VERSION_NUMBER,
+            VERSION_MAP
+        )
+
+        val collection = DatabaseData.Collection(
+            values = setOf(
+                ReferenceWithVersion(
+                    Reference("entity1", backingKey1, VersionMap("entity1" to 1)),
+                    VersionMap("actor" to 1)
+                )
+            ),
+            schema = PRE_UPGRADE_CONTAINER_SCHEMA,
+            databaseVersion = 1,
+            versionMap = VersionMap("actor" to 1)
+        )
+
+        database.insertOrUpdate(backingKey1, entity1)
+        database.insertOrUpdate(collectionKey, collection)
+
+        val outputCollection = database.getCollection(collectionKey, PRE_UPGRADE_CONTAINER_SCHEMA)
+        assertThat(outputCollection!!.values.map {
+            database.getEntity(it.reference.storageKey, PRE_UPGRADE_CONTAINER_SCHEMA)
+        }).containsExactly(entity1)
+
+        // Note that in this case the returned inline entity doesn't contain a null
+        // value for the new field number; though by the time it's returned it does
+        // report as being of the new schema type.
+        val outputVersionOfEntity1 = DatabaseData.Entity(
```

Comment:
can you use 
val outputVersionOfEntity1 = entity1.copy(schema=POST_UPGRADE_CONTAINER_SCHEMA)
to make it easier to see what changed